src/promised-singleton.ts

```diff
@@ -0,0 +1,64 @@
+/**
+ * The PromisedSingleton is a generic wrapper for an asynchronous object that you only
+ * want one instance of (a singleton).
+ *
+ * For the braintree libraries, we only want to load them once. This presents a problem
+ * when several consumers request the clients at once. The Promises don't automatically
+ * chain so you have to do some gatekeeping to make sure only one instance gets created.
+ *
+ * `PromisedSingleton` ensures that no matter how many callers request the object,
+ * only one instance gets created.
+ *
+ * It gets initialized with a Promise that generates the singleton and when `get()` is
+ * first called, it executes the Promise, caches its results and returns it to the caller.
+ *
+ * If more callers call it in the interim, it chains the promises and when the singleton
+ * is created, it resolves them all.
+ *
+ * @export
+ * @class PromisedSingleton
+ * @template T
+ */
+export class PromisedSingleton<T> {
+  /**
+   * Request the singleton
+   *
+   * @returns {Promise<T>}
+   * @memberof PromisedSingleton
+   */
+  async get(): Promise<T> {
+    // if it's already in the cache return it
+    if (this.cachedResponse) {
+      return this.cachedResponse;
+    }
+
+    // if another promise is in line, chain it
+    if (this.previousPromise) {
+      this.previousPromise = this.previousPromise.then(response => {
```

Comment:
We should add an optional second argument to the `then` method to pass a reject callback so the requestor can take appropriate action if the promise fails.

Reply:
The caller is getting a Promise back so it still acts like a Promise and rejects properly, ie:
```
try {
  const service = await serviceProvider.service.get();
} catch(error) {
  console.error('error loading service', error);
}
```
or 
```
serviceProvider.service.get().then(service => { ... }).catch(error => { ... });
```
Is that what you were saying? I went ahead and added some rejection tests to verify the behavior.

Reply:
Yeah, I like setting up methods that create and return a promise to accept both the success and error callbacks like the `.then` method does, just for feature parity. Chaining works fine.